test/app/App/Panel/AddToken/index.test.js

```diff
@@ -82,23 +82,107 @@ describe('setting token address', () => {
     expect(contractAddressInput.textContent).toBe('')
   })
 
-  it('should update add token navigation when an address is entered', async () => {
+  it('should update add token navigation when an invalid contract address is entered', async () => {
     const { user, getByLabelText, getByRole } = setupComponent(
-      <AddToken data={{ notifyData: { chainId: 137 } }} />
+      <AddToken data={{ notifyData: { chainId: 1 } }} />
     )
+    const contractAddressInput = getByLabelText(`Enter token's address`)
+    await user.type(contractAddressInput, 'INVALID_ADDRESS')
+    const setAddressButton = getByRole('button', { name: 'Set Address' })
+    await user.click(setAddressButton)
+
+    expect(link.send).toHaveBeenCalledTimes(1)
+    expect(link.send).toHaveBeenCalledWith('nav:forward', 'dash', {
+      view: 'tokens',
+      data: {
+        notify: 'addToken',
+        notifyData: {
+          chainId: 1,
+          address: 'INVALID_ADDRESS',
+          error: 'INVALID CONTRACT ADDRESS',
+          tokenData: undefined
+        }
+      }
+    })
+  })
+
+  it('should update add token navigation when a contracts details cannot be validated on-chain', async () => {
+    link.invoke.mockImplementationOnce((action, address, chainId) => {
+      expect(action).toBe('tray:getTokenDetails')
+      expect(address).toBe('0x3432b6a60d23ca0dfca7761b7ab56459d9c964d0')
+      expect(chainId).toBe(1)
+      return {
+        decimals: 0,
+        name: '',
+        symbol: '',
+        totalSupply: ''
+      }
+    })
 
+    const { user, getByLabelText, getByRole } = setupComponent(
+      <AddToken data={{ notifyData: { chainId: 1 } }} />
+    )
     const contractAddressLabel = getByLabelText(`Enter token's address`)
     await user.type(contractAddressLabel, '0x3432b6a60d23ca0dfca7761b7ab56459d9c964d0')
     const setAddressButton = getByRole('button', { name: 'Set Address' })
     await user.click(setAddressButton)
 
-    expect(link.send).toHaveBeenCalledWith('tray:action', 'navDash', {
+    expect(link.invoke).toHaveBeenCalledTimes(1)
```

Comment:
minor point but you don't need to verify this since if it wasn't called the test won't pass

Reply:
yep, in some cases (like the line below) I think it makes sense, but in these cases we have to do it like that because the mocked function doesn't return anything -- we're just checking that it was called. for functions that return a value we usually don't need to check it because it if returns the wrong value (or nothing) it should break something